spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/MySqlJdbcInfoProviderAutoConfiguration.java

```diff
@@ -0,0 +1,59 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.sql;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides the {@link CloudSqlJdbcInfoProvider} for MySQL.
+ *
+ * @author João André Martins
+ */
+@Configuration
+@ConditionalOnClass({com.google.cloud.sql.mysql.SocketFactory.class, GcpCloudSqlProperties.class})
```

Comment:
Shound't we be sure in that we have MySQL on CP as well, e.g. `@ConditionalOnClass` for the `com.mysql.jdbc.Driver` class?

Also: shound't we have something like `@AutoConfigureBefore(GcpCloudSqlAutoConfiguration.class)`  ?

Reply:
> Also: shound't we have something like `@AutoConfigureBefore(GcpCloudSqlAutoConfiguration.class)` ?

I also thought about that, but then I'm afraid that `appengineCloudSqlJdbcInfoProvider()` would not be correctly generated, for when an app runs on App Engine.

Reply:
OK. So who has a precedence? `appengineCloudSqlJdbcInfoProvider`? So, this one should be extracted to its own `AutoConfig` class and mustbe marked with the `@AutoConfigureBefore(GcpCloudSqlAutoConfiguration.class, MySqlJdbcInfoProviderAutoConfiguration.class, PostgreSqlJdbcInfoProviderAutoConfiguration.class)`

Reply:
Would adding `@Primary` to `appengineCloudSqlJdbcInfoProvider` also do?
I'd like to prevent adding yet another auto configuration class if I can help it...

Reply:
No, that's not good. The `@Primary` is for use-cases when you have several beans of the same type. It's not related to what we would like to achieve here - single unique `CloudSqlJdbcInfoProvider` bean.

With the current state you don't have guarantee which class will be loaded first. So, you may have `MySqlJdbcInfoProviderAutoConfiguration` configured already, but we do not even go to the `appengineCloudSqlJdbcInfoProvider` to have its `@Primary` on board because we don't overcome  `@ConditionalOnMissingBean(CloudSqlJdbcInfoProvider.class)`.

That's not a big deal to have `AutoConfig` class per feature. What is the obstacle do you see that way?

Reply:
For me, less classes makes the code easier to understand.
But you're right, we could run into that situation. Will fix.